23.0.0.8-beta staging (#3239)
OpenLiberty/blogs: posts/2023-08-03-23.0.0.8-beta.adoc

---
layout: post
title: "Spring Boot 3 support in Open Liberty 23.0.0.8-beta"
# Do NOT change the categories section
categories: blog
author_picture: https://avatars3.githubusercontent.com/mbroz2
author_github: https://github.com/mbroz2
seo-title: Spring Boot 3 support in Open Liberty 23.0.0.8-beta - OpenLiberty.io

> seo-description: Open Liberty 23.0.0.8-beta introduces support for Spring Boot 3, enabling developers to take advantage of utilizing Spring Boot 3.x function within their Liberty applications.  This beta release also includes the very early preview of Jakarta Data.  

Reviewer: ```suggestion
seo-description: Open Liberty 23.0.0.8-beta introduces support for Spring Boot 3, so you can use Spring Boot 3.x function in your Liberty applications. This beta release also includes the very early preview of Jakarta Data. 
```

> blog_description: Open Liberty 23.0.0.8-beta introduces support for Spring Boot 3, enabling developers to take advantage of utilizing Spring Boot 3.x function within their Liberty applications.  This beta release also includes the very early preview of Jakarta Data.  

Reviewer: ```suggestion
blog_description: Open Liberty 23.0.0.8-beta introduces support for Spring Boot 3, so you can use Spring Boot 3.x function in your Liberty applications. This beta release also includes the very early preview of Jakarta Data.  
```

open-graph-image: https://openliberty.io/img/twitter_card.jpg
open-graph-image-alt: Open Liberty Logo
---
= Spring Boot 3 support in Open Liberty 23.0.0.8-beta
Michal Broz <https://github.com/mbroz2>
:imagesdir: /
:url-prefix:
:url-about: /
//Blank line here is necessary before starting the body of the post.

Open Liberty 23.0.0.8-beta introduces support for Spring Boot 3, so you can use Spring Boot 3.x function in your Liberty applications.  This beta release also includes the very early preview of Jakarta Data.  

The link:{url-about}[Open Liberty] 23.0.0.8-beta includes the following beta features (along with link:{url-prefix}/docs/latest/reference/feature/feature-overview.html[all GA features]):

* <<sp3, Spring Boot 3.0 support>>
* <<data, Early preview of Jakarta Data>>

See also link:{url-prefix}/blog/?search=beta&key=tag[previous Open Liberty beta blog posts].

// // // // DO NOT MODIFY THIS COMMENT BLOCK <GHA-BLOG-TOPIC> // // // // 
// Blog issue: https://github.com/OpenLiberty/open-liberty/issues/25679
// Contact/Reviewer: ReeceNana,hlhoots
// // // // // // // // 
[#sp3]
== Spring Boot 3.0 support

> The Liberty Spring Boot Support 3.0 feature provides more complete support for running a Spring Boot 3.0 application on Liberty as well as having the capability to "thin" the application when creating applications in containers. 

Reviewer: ```suggestion
The Liberty Spring Boot Support 3.0 feature provides more complete support for running a Spring Boot 3.0 application on Liberty. It also provides the capability to "thin" the application when you create applications in containers. 
```

> Prior releases of Liberty provided support for Spring Boot 1.5 and Spring Boot 2.0 applications.  In the 23.0.0.3 release, support was extended to include link:/blog/2023/06/15/running-spring-boot-3.html[Running a Spring Boot 3.x application WAR file on Liberty].  With this new Spring Boot 3.0 feature, users now have support for both `.wars` and `.jars` by enabling the `springboot-3.0` feature.

Reviewer 2: Prior releases of Liberty provided support Spring Boot 1.5 and Spring Boot 2.x applications. With the creation of this new Spring Boot 3.0 feature, users will now have support for deploying Spring Boot 3.x applications by enabling the springboot-3.0 feature. Liberty has always been able to support Spring Boot Applications packaged as a war, as demonstrated in this recent blog entry (https://openliberty.io/blog/2023/06/15/running-spring-boot-3.html). This new capability enables both "jar and war" support.

> Prior releases of Liberty provided support for Spring Boot 1.5 and Spring Boot 2.0 applications.  In the 23.0.0.3 release, support was extended to include link:/blog/2023/06/15/running-spring-boot-3.html[Running a Spring Boot 3.x application WAR file on Liberty].  With this new Spring Boot 3.0 feature, users now have support for both `.wars` and `.jars` by enabling the `springboot-3.0` feature.

Reviewer: ```suggestion
Prior releases of Liberty provided support for Spring Boot 1.5 and Spring Boot 2.0 applications You can deploy Spring Boot 3.x applications by enabling the `springboot-3.0` feature. Liberty has always supported Spring Boot Applications packaged as a WAR file, as demonstrated in link:https://openliberty.io/blog/2023/06/15/running-spring-boot-3.html[ this recent blog post]. The Spring Boot 3.0 feature enables both JAR and WAR file support.
```

Reviewer 3: ```suggestion
Prior releases of Liberty provided support for Spring Boot 1.5 and Spring Boot 2.0 applications.  Liberty has also supported Spring Boot Applications packaged as a WAR file, as demonstrated in link:https://openliberty.io/blog/2023/06/15/running-spring-boot-3.html[this recent blog post].  The Open Liberty 23.0.0.8-beta release enables you to deploy Spring Boot 3.x applications by enabling the `springboot-3.0` feature, with support for both JAR and WAR file types.
```

> To run a Spring Boot 3.x application on Liberty, you must be running with Java 17 or higher.  Additionally, if your application uses the link:/docs/latest/reference/feature/servlet.html[Jakarta Servlet] feature, it must be Jakarta Servet 6.0.  The features can be configured via the `server.xml` file as follows:

Reviewer: ```suggestion
To run a Spring Boot 3.x application on Liberty, you must be running with Java 17 or higher.  Additionally, if your application uses the link:/docs/latest/reference/feature/servlet.html[Jakarta Servlet] feature, it must be Jakarta Servet 6.0. Configure these features in the `server.xml` file, as shown in the following example:
```

[source,xml]
----
<features>
   <feature>springboot-3.0</feature>
   <feature>servlet-6.0</feature>
</features>
----

> As with previous versions of the Spring Boot feature, specify link:/docs/latest/reference/config/springBootApplication.html[Spring Boot configuration elements] in the `server.xml` file. For example:

Reviewer 3: ```suggestion
As with previous versions of the Spring Boot Support feature, Spring Boot application JAR files can be placed in the `/dropins/spring/ directory.  Alternatively, specify the link:/docs/latest/reference/config/springBootApplication.html[Spring Boot configuration elements] in the `server.xml` file. For example:
```

[source,xml]
----
 <springBootApplication id="spring-boot-app" location="spring-boot-app-0.1.0.jar" name="spring-boot-app" />
----

> 

Reviewer 2: Similar to previous versions, SB application jar's can use the /dropins/spring folder, and skip server.xml entries.

---
layout: post
title: "Spring Boot 3 support in Open Liberty 23.0.0.8-beta"
# Do NOT change the categories section
categories: blog
author_picture: https://avatars3.githubusercontent.com/mbroz2
author_github: https://github.com/mbroz2
seo-title: Spring Boot 3 support in Open Liberty 23.0.0.8-beta - OpenLiberty.io
seo-description: Open Liberty 23.0.0.8-beta introduces support for Spring Boot 3, enabling developers to take advantage of utilizing Spring Boot 3.x function within their Liberty applications.  This beta release also includes the very early preview of Jakarta Data.  
blog_description: Open Liberty 23.0.0.8-beta introduces support for Spring Boot 3, enabling developers to take advantage of utilizing Spring Boot 3.x function within their Liberty applications.  This beta release also includes the very early preview of Jakarta Data.  
open-graph-image: https://openliberty.io/img/twitter_card.jpg
open-graph-image-alt: Open Liberty Logo
---
= Spring Boot 3 support in Open Liberty 23.0.0.8-beta
Michal Broz <https://github.com/mbroz2>
:imagesdir: /
:url-prefix:
:url-about: /
//Blank line here is necessary before starting the body of the post.

> Open Liberty 23.0.0.8-beta introduces support for Spring Boot 3, enabling developers to take advantage of utilizing Spring Boot 3.x function within their Liberty applications.  This beta release also includes the very early preview of Jakarta Data.  

Reviewer 4: ```suggestion
Open Liberty 23.0.0.8-beta introduces support for Spring Boot 3, so you can use Spring Boot 3.x function in your Liberty applications.  This beta release also includes the very early preview of Jakarta Data.  
```

The link:{url-about}[Open Liberty] 23.0.0.8-beta includes the following beta features (along with link:{url-prefix}/docs/latest/reference/feature/feature-overview.html[all GA features]):

> * <<sp3, Spring Boot 3.0 Support>>

Reviewer 4: ```suggestion
* <<sp3, Spring Boot 3.0 support>>
```

> * <<data, Early Preview of Jakarta Data>>

Reviewer 4: ```suggestion
* <<data, Early preview of Jakarta Data>>
```

See also link:{url-prefix}/blog/?search=beta&key=tag[previous Open Liberty beta blog posts].

// // // // DO NOT MODIFY THIS COMMENT BLOCK <GHA-BLOG-TOPIC> // // // // 
// Blog issue: https://github.com/OpenLiberty/open-liberty/issues/25679
// Contact/Reviewer: ReeceNana,hlhoots
// // // // // // // // 
[#sp3]

> == Spring Boot 3.0 Support

Reviewer 4: ```suggestion
== Spring Boot 3.0 support
```

> The Liberty Spring boot 3.0 feature provides a more complete support for running a Sprint Boot 3.0 application on Liberty as well as having the capability to thin the application when creating applications in containers. 

Reviewer 4: ```suggestion
The Liberty Spring Boot Support 3.0 feature provides more complete support for running a Spring Boot 3.0 application on Liberty as well as having the capability to "thin" the application when creating applications in containers. 
```

> Prior releases of Liberty provided support for Spring Boot 1.5 and Spring Boot 2.0 applications.  In the 23.0.0.3 release, support was extended to include link:/blog/2023/06/15/running-spring-boot-3.html[Running a Spring Boot 3.x application WAR file on Liberty].  With the creation of this new Spring Boot 3.0 feature, users will now have support for both `.wars` and `.jars` by enabling the `springboot-3.0` feature.

Reviewer 4: ```suggestion
Prior releases of Liberty provided support for Spring Boot 1.5 and Spring Boot 2.0 applications.  In the 23.0.0.3 release, support was extended to include link:/blog/2023/06/15/running-spring-boot-3.html[Running a Spring Boot 3.x application WAR file on Liberty].  With this new Spring Boot 3.0 feature, users now have support for both `.wars` and `.jars` by enabling the `springboot-3.0` feature.
```

> To utilize the Spring Boot 3.x functionality, you must be running with Java 17 or higher.  Additionally, if your application makes use of the link:/docs/latest/reference/feature/servlet.html[Jakarta Servlet] feature, it must be version 6.0.  These features would be configured in the `server.xml` file as follows:

Reviewer 4: ```suggestion
To run a Spring Boot 3.x application on Liberty, you must be running with Java 17 or higher.  Additionally, if your application uses the link:/docs/latest/reference/feature/servlet.html[Jakarta Servlet] feature, it must be Jakarta Servet 6.0.  The features can be configured via the `server.xml` file as follows:
```

[source,xml]
----
<features>
   <feature>springboot-3.0</feature>
   <feature>servlet-6.0</feature>
</features>
----

> Additionally, similarly to previous feature versions, link:/docs/latest/reference/config/springBootApplication.html[Spring Boot configuration elements] can be specified via the `server.xml` file. For example:

Reviewer 4: ```suggestion
As with previous versions of the Spring Boot feature, specify link:/docs/latest/reference/config/springBootApplication.html[Spring Boot configuration elements] in the `server.xml` file. For example:
```

[source,xml]
----
 <springBootApplication id="spring-boot-app" location="spring-boot-app-0.1.0.jar" name="spring-boot-app" />
----

> A few migration notes for using Spring Boot 3:

Reviewer 4: ```suggestion
If you have used previous versions of the Spring Boot feature and are now migrating your application to use Spring Boot 3, note the following requirements:
```

* link:/docs/latest/reference/feature/webProfile-10.0.html[Jakarta EE 10 features] must be used. 

* Security configuration requires in-application modifications. 
  - link:https://spring.io/blog/2022/02/21/spring-security-without-the-websecurityconfigureradapter[Spring Security without the WebSecurityConfigurerAdapter]

  - link:https://docs.spring.io/spring-security/reference/servlet/authorization/authorize-http-requests.html[Authorize HttpServletRequests]

>   - link:https://stackoverflow.com/questions/72381114/spring-security-upgrading-the-deprecated-websecurityconfigureradapter-in-spring[Spring Security: Upgrading the deprecated WebSecurityConfigurerAdapter in Spring Boot 2.7.0]

Reviewer 3: @cbridgha @hlhoots how necessary is this? Is the content available elsewhere?  
I don't particularly like linking to an answer on stackoverflow in a blog post since the content there can be 'easily' altered to contain anything.

Reviewer 3: I just checked, and there's no precedence to linking to a stackoverflow answer.

> 
>   - link:https://stackoverflow.com/questions/72381114/spring-security-upgrading-the-deprecated-websecurityconfigureradapter-in-spring[Spring Security: Upgrading the deprecated WebSecurityConfigurerAdapter in Spring Boot 2.7.0]

Reviewer 3: ```suggestion
```

> * A new server template (`springBoot3`) is available.

Reviewer 4: Sorry, I don't know how the Spring Boot support works so this may be a daft question. What does the "server template" refer to? Liberty server or a Spring Boot concept? Should this link to something or is it patently obvious to a Spring Boot developer where to find it? Or would a Liberty developer who's not use to using either Spring Boot or the Liberty Spring Boot feature, not know what this refers to?

Reviewer 3: the template refers to template server.xml files that we ship with the product.  I was going to link to documentation on this here, but we didn't have any, so I opened an issue for it last night: https://github.com/OpenLiberty/docs/issues/6806

Reviewer: ^^ we'll add this with 23.0.0.8

   
// DO NOT MODIFY THIS LINE. </GHA-BLOG-TOPIC> 

[#data]

> == Early Preview of Jakarta Data

Reviewer 4: ```suggestion
== Early preview of Jakarta Data
```

> The 23.0.0.8-beta release also includes a preview of Jakarta Data.  More information is available in the link:{url-prefix}/blog/2023/02/21/23.0.0.7-beta.html[Jakarta Data beta 2 update in Open Liberty 23.0.0.7-beta] blog post.

Reviewer 4: ```suggestion
The 23.0.0.8-beta release also includes a preview of Jakarta Data.  More information is available in the link:{url-prefix}/blog/2023/07/11/23.0.0.7-beta.html[Jakarta Data beta 2 update in Open Liberty 23.0.0.7-beta] blog post.
```

> 

Reviewer 4: I like this (providing a link back to the info in a previous blog post).

[#run]
=== Try it now 

To try out these features, update your build tools to pull the Open Liberty All Beta Features package instead of the main release. The beta works with Java SE 20, Java SE 17, Java SE 11, and Java SE 8.

> If you're using link:{url-prefix}/guides/maven-intro.html[Maven], you can install the All Beta Features package using:

Reviewer 4: ```suggestion
If you're using link:{url-prefix}/guides/maven-intro.html[Maven], add the All Beta Features package to your `pom.xml` file:
```

Reviewer 3: Note that all wording from this one down are coming from the template.  So if we want different wording, it should be updated in the template for future posts.

[source,xml]
----
<plugin>
    <groupId>io.openliberty.tools</groupId>
    <artifactId>liberty-maven-plugin</artifactId>
    <version>RELEASE</version>
    <configuration>
        <runtimeArtifact>
          <groupId>io.openliberty.beta</groupId>
          <artifactId>openliberty-runtime</artifactId>
          <version>23.0.0.8-beta</version>
          <type>zip</type>
        </runtimeArtifact>
    </configuration>
</plugin>
----

> You must also add dependencies to your pom.xml file for the beta version of the APIs that are associated with the beta features that you want to try.  For example, for Jakarta Data Beta 2, you would include:

Reviewer 4: How does one find out what Maven dependencies for the beta APIs are needed? Aside from the Jakarta Data one that's given as an example, are they listed anywhere?

Reviewer 3: Usually the blog post topic that requires it will provide the necessary information.

[source,xml]
----
<dependency>
  <groupId>jakarta.data</groupId>
  <artifactId>jakarta-data-api</artifactId>
  <version>1.0.0-b2</version>
</dependency>
----

> Or for link:{url-prefix}/guides/gradle-intro.html[Gradle]:

Reviewer 4: Do you need to add dependencies of the beta APIs in Gradle too?

Reviewer 4: Not sure the "Or..." format of the lead-in sentences works so well when there is the additional bit after the Maven paragraph about adding beta API dependencies too. It's not so clear what the "Or" refers back to (actually the "Maven....All Beta Features" sentence).

Reviewer 3: it's suppose to refer to the maven.

[source,gradle]
----
dependencies {
    libertyRuntime group: 'io.openliberty.beta', name: 'openliberty-runtime', version: '[23.0.0.8-beta,)'
}
----

Or if you're using link:{url-prefix}/docs/latest/container-images.html[container images]:

[source]
----
FROM icr.io/appcafe/open-liberty:beta
----

> Or take a look at our link:{url-prefix}/downloads/#runtime_betas[Downloads page].

Reviewer 4: In what scenarios do we still expect people to want the downloadable beta package?

Reviewer 3: One example would be that they don't use maven nor gradle
